storages/backends/gcloud.py

```diff
@@ -0,0 +1,233 @@
+    credentials = setting('GS_CREDENTIALS', None)
+    bucket_name = setting('GS_BUCKET_NAME', None)
+    auto_create_bucket = setting('GS_AUTO_CREATE_BUCKET', False)
+    default_acl = setting('GS_DEFAULT_ACL', 'public-read')
+    bucket_acl = setting('GS_BUCKET_ACL', default_acl)
+    file_name_charset = setting('GS_FILE_NAME_CHARSET', 'utf-8')
+    file_overwrite = setting('GS_FILE_OVERWRITE', True)
+    # The max amount of memory a returned file can take up before being
+    # rolled over into a temporary file on disk. Default is 0: Do not roll over.
+    max_memory_size = setting('GS_MAX_MEMORY_SIZE', 0)
+
+    def __init__(self, **settings):
+        # check if some of the settings we've provided as class attributes
+        # need to be overwritten with values passed in here
+        for name, value in settings.items():
+            if hasattr(self, name):
+                setattr(self, name, value)
+
+        self._bucket = None
+        self._client = None
+
+    @property
+    def client(self):
+        if self._client is None:
+            self._client = Client(
+                project=self.project_id,
+                credentials=self.credentials
+            )
+        return self._client
+
+    @property
+    def bucket(self):
+        if self._bucket is None:
+            self._bucket = self._get_or_create_bucket(self.bucket_name)
+        return self._bucket
+
+    def _get_or_create_bucket(self, name):
+        """
+        Retrieves a bucket if it exists, otherwise creates it.
+        """
+        try:
+            return self.client.get_bucket(name)
+        except NotFound:
+            if self.auto_create_bucket:
+                bucket = self.client.create_bucket(name)
+                bucket.acl.all().grant(self.bucket_acl)
+                bucket.acl.save()
+                return bucket
+            raise ImproperlyConfigured("Bucket %s does not exist. Buckets "
+                                       "can be automatically created by "
+                                       "setting GS_AUTO_CREATE_BUCKET to "
+                                       "``True``." % name)
+
+    def _normalize_name(self, name):
+        """
+        Normalizes the name so that paths like /path/to/ignored/../something.txt
+        and ./file.txt work.  Note that clean_name adds ./ to some paths so
+        they need to be fixed here.
+        """
+        return safe_join('', name)
+
+    def _encode_name(self, name):
+        return smart_str(name, encoding=self.file_name_charset)
+
+    def _open(self, name, mode='rb'):
+        name = self._normalize_name(clean_name(name))
+        file_object = GoogleCloudFile(name, mode, self)
+        if not file_object.blob:
+            raise IOError(u'File does not exist: %s' % name)
+        return file_object
+
+    def _save(self, name, content):
+        cleaned_name = clean_name(name)
+        name = self._normalize_name(cleaned_name)
+
+        content.name = cleaned_name
```

Comment:
This is for compatibility with `s3boto`. My basic approach is that someone should be able to take code that works with `s3boto`, drop in `gcloud`, and have it work (including quirks like this).

We do run `_encode_name` before creating the remote object, but the unencoded name is passed to `upload_from_file` (in `content`), similarly to how `s3boto` passes the unencoded name to `set_contents_from_file`.